analyses/batch-effects/graph-generator.Rmd

```diff
@@ -0,0 +1,426 @@
+# Batch Effects Graphs Description
+
+This markdown file also exists as a jupyter notebook file for a cleaner version.
+# Library Import & Data Download
+
+#### Import the necessary libararies 
+
+
+```{R}
+library(tibble)
+library(tidyr)
+library(ggplot2)
+library(readr)
+library(dplyr)
+library(BatchQC)
+library(forcats)
+library(sva)
+library(rprojroot)
+```
+
+####  Configure file paths & import functions
+
+
+```{R}
+# Configure file paths
+root_dir = find_root(has_file("OpenPBTA-analysis.Rproj"))
+analysis_dir = file.path(root_dir, "analyses", "batch-effects")
+data_dir = file.path(root_dir, "data")
+functions = file.path(analysis_dir, "util", "functions.R")
+```
+
+
+```{R}
+# Load functions from /util/function.R
+source(functions)
+```
+
+#### Download the Data
+
+
+```{R}
+# Download all covariate data which will be used to identify batches
+covariate_file = file.path(data_dir, "pbta-histologies.tsv")
+covariate = read_tsv(covariate_file, col_types = cols(molecular_subtype = "c"))
+```
+
+
+```{R}
+# Download gene expression data
+dat_rsem_polya_file = file.path(data_dir, "pbta-gene-expression-rsem-tpm.polya.rds")
+dat_rsem_polya = readRDS(dat_rsem_polya_file)
+
+dat_rsem_stranded_file = file.path(data_dir, "pbta-gene-expression-rsem-tpm.stranded.rds")
+dat_rsem_stranded = readRDS(dat_rsem_stranded_file)
+
+dat_kallisto_stranded_file = file.path(data_dir, "pbta-gene-expression-kallisto.stranded.rds")
+dat_kallisto_stranded <- readRDS(dat_kallisto_stranded_file)
+dat_kallisto_stranded = dat_kallisto_stranded[,2:ncol(dat_kallisto_stranded)]
+
+dat_kallisto_polya_file = file.path(data_dir, "pbta-gene-expression-kallisto.polya.rds")
+dat_kallisto_polya <- readRDS(dat_kallisto_polya_file)
+```
+
+##### Shorten the data files if you don't have a strong enough computer 
+
+***This should not be run for a full analysis!***
+
+
+```{R}
+# Shorten Data Files
+dat_rsem_polya = shorten(dat_rsem_polya)
+dat_rsem_stranded = shorten(dat_rsem_stranded)
+dat_kallisto_stranded = shorten(dat_kallisto_stranded)
+dat_kallisto_polya = shorten(dat_kallisto_polya)
+```
```

Comment:
I don't think you want to subset to the first 100 lines, which is what `shorten()` does, by _default_. Instead, I'd recommend making this optional, where you do **not** do it by default. There is a limit to the amount of RAM that the instance running CI has (I _believe_ it's 8 GB), so it could come in handy for that, too.

My first question: Do you have an idea of how much memory is required to run the analysis without the `shorten()` step?

If it is likely to be too costly in terms of memory, we have some options.

If you're going to use R Markdown, the way I'd recommend doing this is with [`params`](https://rmarkdown.rstudio.com/lesson-6.html).

In the YAML header of the file, you'd add:

```yaml
params:
  shorten: FALSE
```

And then in this chunk you could do something like:

```R
if (params$shorten) {
  dat_rsem_polya = shorten(dat_rsem_polya)
  dat_rsem_stranded = shorten(dat_rsem_stranded)
  dat_kallisto_stranded = shorten(dat_kallisto_stranded)
  dat_kallisto_polya = shorten(dat_kallisto_polya)
}
```

If you wanted to run the R Markdown document from the command line and run the shorten step, you can do the following:

```sh
Rscript -e "rmarkdown::render('analyses/batch-effects/graph-generator.Rmd', params = list(shorten = TRUE), clean = TRUE)"
```

These chunks are untested, but are hopefully enough to give you an idea of how to accomplish this!

Another way to "shorten" the data frame that is a little more principled (at least for plotting PCs) would be to subset to genes with high variance. You can use [`matrixStats::rowVars()`](https://www.rdocumentation.org/packages/matrixStats/versions/0.58.0/topics/rowVars) for calculating the variance and if I recall correctly, that calculation probably won't be prohibitive in terms of memory or time.

Reply:
This is a great idea. I'll work on integrating the changes. I'm not sure the exact amount of memory that this requires. But I can get back to you on that.

Reply:
I was doing that so I could run this analysis on my machine that has eight GB. It would be great if you could run this on a server that has 32 GB. Then I could just delete the code that does this trimming. Would that work?

Reply:
We will still want to run it in continuous integration, which has a cap of 8 GB of RAM as well.

Reply:
I'm resending the graphs from the full dataset. I can generate these graphs when I run them on a server with more memory

Reply:
The `matrixStats::rowVars()` method seems to work well until I get to the third "batch" where I'm combining polyA and rsem files together. If I select rows based on which ones have the most variance, I end up with a different subset of genes in the two different gene expression files. As such, I'm not able to join the two files together and do any batch correction.

This isn't a problem if I run my full analysis (without shortening files) on a big server with a lot of memory or if I do a simplified file shortening method for this last batch as I did earlier:
`df=df[1:500,]`

To your point though, we don't want to draw conclusions on a few genes alone. I could try seeing what's the largest amount of genes that I could include without hitting the 8GB memory limit.